Refactor rtoptions: Centralize environment variable handling

msudumbrekar-TT opened the pull request:
Supersedes Previous PR : https://github.com/tenstorrent/tt-metal/pull/30803
# Consolidate Environment Variable Parsing

## Overview
Consolidated scattered and redundant `getenv()` calls into a centralized switch-based system with inline documentation for all environment variables.

## Problem
- Redundant `getenv()` calls scattered across multiple functions
- Duplicate parsing logic
- No central documentation

## Solution
- **EnvVarID Enum**: Type-safe enumeration of all variables
- **HandleEnvVar()**: Single switch statement with all parsing logic
- **Inline Documentation**: Description, defaults, and examples for each variable

## Changes
- Consolidated 70+ environment variables into one location
- Removed redundant initialization code
- Applied code formatting fixes

## Testing
 All tests passing locally (device init, device pool, hardware validation)

**CI Workflows:**
- All post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/19307654899
- BH post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/19307647345

## Files Modified
- `tt_metal/llrt/rtoptions.cpp` 
- `tt_metal/llrt/rtoptions.hpp` 

## Benefits
- Single source of truth for all environment variables
- Eliminated code duplication
- Self-documenting with inline comments
- Easier to maintain and extend

msudumbrekar-TT committed:
Refactor rtoptions: Centralize environment variable handling and remove duplicates

- Migrate to enchantum-based EnvVarID enum for type-safe environment variable handling
- Implement table-driven environment variable parsing via ENV_VAR_TABLE
- Remove ~150 lines of duplicate manual environment variable parsing from constructor
- Centralize all env var logic in HandleEnvVar() switch statement
- Optimize profiler dependency checks to use this->profiler_enabled instead of redundant getenv() calls
- Add comprehensive documentation for all environment variables
- Add TT_METAL_HOME fallback for backwards compatibility
- Fix Tracy preprocessor blocks and constructor initialization order

This consolidates and cleans up environment variable handling throughout rtoptions,
making it more maintainable and reducing code duplication.

Supersedes: #<old_PR_number> (will add link)

blozano-tt commented on tt_metal/llrt/rtoptions.cpp:
There are a lot of useful comments here. I wonder if you could switch to doxygen or rst format so they can be included in autogenerated documentation?

msudumbrekar-TT committed:
Address GitHub review comments and add helper functions

- Fix use_mesh_graph_descriptor_2_0 redundant logic (make concise)
- Remove TT_METAL_CORE_GRID_OVERRIDE_TODEPRECATE_ENV_VAR, use enchantum::to_string
- Remove empty comment line in rtoptions.hpp
- Add is_env_enabled() helper to replace 14 instances of value[0] == '1'
- Fix ::tolower undefined behavior with lambda using unsigned char
- Fix catch(...) to catch specific exceptions (std::invalid_argument, std::out_of_range)
- Fix syntax error: missing :: in EnvVarID scope resolution

msudumbrekar-TT committed:
Remove TT_METAL_HOME fallback per review feedback

Main branch only parses TT_METAL_RUNTIME_ROOT. Tests work via current_path fallback.

blozano-tt commented on tt_metal/llrt/rtoptions.cpp:
```diff
@@ -32,15 +33,31 @@ const char* RunTimeDebugFeatureNames[RunTimeDebugFeatureCount] = {
+}
+
+// Helper function to check if environment variable value is "1" (enabled)
+inline bool is_env_enabled(const char* value) { return value && value[0] == '1'; }
```
I don't think `inline` is needed. Compiler will do what it wants.

```suggestion
bool is_env_enabled(const char* value) { return value && value[0] == '1'; }
```

msudumbrekar-TT committed:
Remove inline keyword

msudumbrekar-TT committed:
Add TT_METAL_FORCE_JIT_COMPILE environment variable support

ruizhangTT commented on tt_metal/llrt/rtoptions.cpp:
```diff
@@ -89,314 +106,840 @@ RunTimeOptions::RunTimeOptions() {
-}
+        // Only process if the environment variable is set
+        if (value) {
+            HandleEnvVar(id, value);
```
I see some branches in `HandleEnvVar` also tests whether `value == nullptr`.  If we only call it when `value != nullptr`, then we don't need to test inside?  Or we could always test inside and not test here (might have the benefit of not needing line 930-938 below).

ruizhangTT commented on tt_metal/llrt/rtoptions.cpp:
```diff
@@ -89,314 +106,840 @@ RunTimeOptions::RunTimeOptions() {
+void RunTimeOptions::InitializeFromEnvVars() {
+    // Use enchantum to automatically iterate over all EnvVarID enum values
+    for (const auto [id, name] : enchantum::entries_generator<EnvVarID>) {
+        const char* value = std::getenv(std::string(name).c_str());
```
minor nit but might be ok to use `name.data()` since entries_generator generates null-terminated strings.  https://github.com/ZXShady/enchantum/blob/main/docs/features.md#entries_generator

msudumbrekar-TT replied:
Done!

ruizhangTT commented on tt_metal/llrt/rtoptions.hpp:
```diff
@@ -30,6 +28,128 @@
 namespace llrt {
+// ============================================================================
+// ENVIRONMENT VARIABLE IDs
+enum class EnvVarID {
```
nit: might be ok to only have a fwd decl here and put the def in the `.cpp` file.  This potentially reduces recompilation when adding new env vars.

msudumbrekar-TT replied:
Done!

ruizhangTT commented on tt_metal/llrt/rtoptions.hpp:
```diff
@@ -57,6 +177,7 @@ enum RunTimeDebugClass {
 
 extern const char* RunTimeDebugFeatureNames[RunTimeDebugFeatureCount];
 extern const char* RunTimeDebugClassNames[RunTimeDebugClassCount];
+class RunTimeOptions;
```
Why was this added?

msudumbrekar-TT replied:
Might have missed removing that! Will remove.

ruizhangTT commented on tt_metal/llrt/rtoptions.cpp:
```diff
@@ -89,314 +106,840 @@ RunTimeOptions::RunTimeOptions() {
+// Central handler that processes environment variables based on their ID
+// Uses switch statement for efficient dispatch
+
+void RunTimeOptions::HandleEnvVar(EnvVarID id, const char* value) {
```
You are right.  I guess the class already has inconsistency on it's own.  Never mind.

msudumbrekar-TT committed:
Address final review feedback

msudumbrekar-TT committed:
Address GitHub review clang tidy comments

msudumbrekar-TT committed:
Remove redundant watcher_disabled_features initialization

msudumbrekar-TT committed:
Remove MGD 1.0 support (deprecated in favor of MGD 2.0)

msudumbrekar-TT committed:
Fix clang-tidy warning: ensure null-terminated string for getenv

msudumbrekar-TT committed:
Update 2-erisc mode to match PR #31833: invert env var and disable for simulators
